tests/verify.rs

```diff
@@ -1,8 +1,170 @@
+                    time,
+                    KeyUsage::server_auth(),
+                    &[],
+                );
+            assert!(matches!(result, Err(Error::NameConstraintViolation)));
+        }
+    }
+}
+
+struct ConstraintTest {
+    trust_anchor: Vec<u8>,
+    permitted_certs: Vec<Vec<u8>>,
+    forbidden_certs: Vec<Vec<u8>>,
+}
+
+fn name_constraints_testcase(webpki_name_constraints: &[u8]) -> ConstraintTest {
+    // Create a trust anchor CA certificate that has the name constraints we want to test.
+    let mut trust_anchor = CertificateParams::new([]);
+    trust_anchor
+        .distinguished_name
+        .push(DnType::CommonName, "Name Constraint Test CA");
+    trust_anchor.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
+    trust_anchor.key_usages = vec![
+        KeyUsagePurpose::KeyCertSign,
+        KeyUsagePurpose::DigitalSignature,
+    ];
+    let name_constraints = rcgen_name_constraints(webpki_name_constraints);
+    trust_anchor.name_constraints = Some(name_constraints.clone());
+    let trust_anchor = Certificate::from_params(trust_anchor).unwrap();
+
+    fn certs_for_subtrees<'a>(
+        trust_anchor: &Certificate,
+        subtrees: impl Iterator<Item = &'a GeneralSubtree>,
+        name_fn: impl Fn(&str) -> String,
+    ) -> Vec<Vec<u8>> {
+        subtrees
+            .filter_map(|subtree| match subtree {
+                GeneralSubtree::DnsName(dns_name) => {
+                    Some(rcgen_ee_for_name(name_fn(dns_name), &trust_anchor))
+                }
+                _ => None,
+            })
+            .collect()
+    }
+
+    // For each permitted subtree in the name constraints, issue an end entity certificate
+    // that contains a DNS name matching the permitted subtree base.
+    let permitted_certs = certs_for_subtrees(
+        &trust_anchor,
+        name_constraints.permitted_subtrees.iter(),
+        |dns_name| format!("valid{}", dns_name).to_string(),
+    );
+
+    // For each permitted subtree in the name constraints, issue an end entity certificate
+    // that contains a DNS name that will **not** match the permitted subtree base.
+    let forbidden_certs = certs_for_subtrees(
+        &trust_anchor,
+        name_constraints.permitted_subtrees.iter(),
+        |dns_name| format!("valid{}.invalid", dns_name).to_string(),
+    );
+
+    ConstraintTest {
+        trust_anchor: trust_anchor.serialize_der().unwrap(),
+        permitted_certs,
+        forbidden_certs,
+    }
+}
+
+fn rcgen_ee_for_name(name: String, issuer: &Certificate) -> Vec<u8> {
+    let mut ee = CertificateParams::new(vec![name.clone()]);
+    ee.distinguished_name.push(DnType::CommonName, name);
+    ee.is_ca = IsCa::NoCa;
+    Certificate::from_params(ee)
+        .unwrap()
+        .serialize_der_with_signer(&issuer)
+        .unwrap()
+}
+
+/// Convert the webpki trust anchor DER encoding of name constraints to rcgen NameConstraints.
+fn rcgen_name_constraints(der: &[u8]) -> NameConstraints {
```

Comment:
Given that the `NameConstraints` symbol only appears here and in the return expression, maybe avoid importing it and stick with `rcgen::NameConstraints` to avoid confusion?

Reply:
Good call, updated.